### PR TITLE
SONARJS-744 Support class properties

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

**Comment by ivandalbosco** [comment]
@vilchik-elena I guess we should color the "static" keyword. I don't know if this is part of this ticket or not.

File: javascript-frontend/src/test/java/org/sonar/javascript/parser/expressions/ClassElementTest.java [hunk 1]
```diff
@@ -35,4 +35,24 @@ public void ok() {
...
+      .matches("1 ;")
+      // string literal
+      .matches("'prop_name' ;")
+      // computed property
+      .matches("[21 + 21] ;")
+
+      .matches("property = 1 ;")
+      .matches("property = foo() ;")
+      .matches("static property ;")
+      .matches("static property = 1 ;")
```
Comment (ivandalbosco) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/expressions/ClassElementTest.java]: @vilchik-elena Hoiw about adding

Comment (ivandalbosco) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/expressions/ClassElementTest.java]: .matches("property1 = property2 = 0");
.matches("property = () => console.log('hello')");

File: javascript-frontend/src/test/java/org/sonar/javascript/parser/expressions/ClassElementTest.java [hunk 2]
```diff
@@ -35,4 +35,24 @@ public void ok() {
...
+
+      .notMatches("property = 1")
+    ;
```
Comment (ivandalbosco) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/expressions/ClassElementTest.java]: How about adding:

Comment (ivandalbosco) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/expressions/ClassElementTest.java]: @vilchik-elena  
.notMatches("property = ;")
.notMatches("static;")

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java [hunk 3]
```diff
@@ -1547,9 +1548,19 @@ public Tree CLASS_ELEMENT() {
...
+          CLASS_FIELD_INITIALIZER(),
           b.token(JavaScriptPunctuator.SEMI)));
   }
 
+  public FieldDeclarationTree CLASS_FIELD_INITIALIZER() {
+    return b.<FieldDeclarationTree>nonterminal()
+      .is(f.fieldDeclaration(
+        b.optional(b.token(JavaScriptLegacyGrammar.STATIC)),
+        PROPERTY_NAME(),
+        b.optional(f.newTuple58(b.token(JavaScriptPunctuator.EQU), ASSIGNMENT_EXPRESSION())),
```
Comment (ivandalbosco) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: @vilchik-elena 
“static” is a reserved word in JavaScript. Don't you think we should prevent the definition of a class property
    static static = 1;
which your implementation currently allows?
BTW, the above instruction, used outside of a class, is currently accepted by our parser. It probably shouldn't be.

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: @ivandalbosco `static` is not reserved word in JS (see http://www.ecma-international.org/ecma-262/6.0/index.html#sec-reserved-words) 
You can right the following code in the same file:
```
var static = 1;
class A {
  static foo(){}
}
```

Moreover this code is valid
```
class C { static static() {}}

class A {
  static static = 1;
}
```

"BTW, the above instruction, used outside of a class, is currently accepted by our parser" - are you sure? it's fails for me.

**Comment by vilchik-elena** [comment]
@ivandalbosco Nice idea about highlighting. Done in 0f58432d2568dd81c23d57b62d6cedb3117a95e4

File: javascript-frontend/src/test/java/org/sonar/javascript/parser/expressions/ClassElementTest.java (hunk 2 above)
Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/expressions/ClassElementTest.java]: @ivandalbosco done in de0236e8b745841e72c5e6709b2529d5e683637c

File: javascript-frontend/src/test/java/org/sonar/javascript/parser/expressions/ClassElementTest.java (hunk 1 above)
Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/expressions/ClassElementTest.java]: @ivandalbosco done in de0236e8b745841e72c5e6709b2529d5e683637c

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FieldDeclarationTree.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ *  Interface for class fields and static properties.
+ *  This syntax is in proposal currently. See https://github.com/jeffmo/es-class-fields-and-static-properties
+ */
+@Beta
+public interface FieldDeclarationTree extends Tree {
+
+  @Nullable
+  SyntaxToken staticToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FieldDeclarationTree.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FieldDeclarationTree.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  Tree propertyName();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FieldDeclarationTree.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FieldDeclarationTree.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+  @Nullable
+  SyntaxToken equalToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FieldDeclarationTree.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FieldDeclarationTree.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+  @Nullable
+  ExpressionTree initializer();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FieldDeclarationTree.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FieldDeclarationTree.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+  @Nullable
+  SyntaxToken semicolonToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/FieldDeclarationTree.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitor.java [hunk 9]
```diff
@@ -219,6 +220,9 @@ public void visitMethodDeclaration(MethodDeclarationTree tree) {
     scanChildren(tree);
   }
 
+  public void visitFieldDeclaration(FieldDeclarationTree tree) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitor.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* MINOR 6 minor

Watch the comments in this conversation to review them.

**Pull request merged**